hera_cal/reflections.py

```diff
@@ -0,0 +1,516 @@
+
+    # locate amplitude peak
+    abs_dfft = np.abs(dfft)
+    abs_dfft_selection = abs_dfft[:, select]
+    ref_peaks = np.max(abs_dfft_selection, axis=1, keepdims=True)
+    ref_dly_inds = np.argmin(np.abs(abs_dfft_selection - ref_peaks), axis=1) + select.min()
+    ref_dlys = dlys[ref_dly_inds, None]
+
+    # calculate shifted peak for sub-bin resolution
+    # https://ccrma.stanford.edu/~jos/sasp/Quadratic_Interpolation_Spectral_Peaks.html
+    # alpha = a, beta = b, gamma = g
+    a = abs_dfft[np.arange(Ntimes), ref_dly_inds-1, None]
+    g = abs_dfft[np.arange(Ntimes), ref_dly_inds+1, None]
+    b = abs_dfft[np.arange(Ntimes), ref_dly_inds, None]
+    bin_shifts = 0.5 * (a - g) / (a - 2 * b + g)
+    
+    # update delay center and peak value given shifts
+    ref_dlys += bin_shifts * np.median(np.diff(dlys))
+    ref_peaks = b - 0.25 * (a - g) * bin_shifts
+    
+    # get reflection significance, defined as ref_peak / median_abs
+    ref_significance = ref_peaks / np.median(abs_dfft_selection, axis=1, keepdims=True)
+
+    # get reflection amplitude
+    ref_amps = ref_peaks / np.max(abs_dfft[:, np.abs(dlys) < np.abs(dly_range).min()], axis=1, keepdims=True)
+
+    # get reflection phase by fitting cosines to filtered data thinned along frequency axis
+    filt = np.zeros_like(dfft)
+    filt[:, select] = dfft[:, select]
+    if Nfreqs % 2 == 1:
+        filt[:, -select] = dfft[:, -select]
+    else:
+        filt[:, -select-1] = dfft[:, -select-1]
+    filt = np.fft.ifft(filt, axis=1)
+    if zero_pad > 0:
+        filt = filt[:, zero_pad:-zero_pad]
+        freqs = freqs[zero_pad:-zero_pad]
+        t = t[:, zero_pad:-zero_pad]
+    phases = np.linspace(0, 2*np.pi, 500, endpoint=False)
+    cosines = np.array([construct_reflection(freqs[::fthin], ref_amps, ref_dlys, p) for p in phases])
+    residuals = np.sum((filt[None, :, ::fthin].real - cosines.real * t[None, :, ::fthin])**2, axis=-1)
+    ref_phs = phases[np.argmin(residuals, axis=0)][:, None] % (2*np.pi)
+    dlys = np.fft.fftfreq(len(freqs), d=dnu) * 1e9
+
+    # get reflection phase by interpolation: this didn't work well in practice
+    # when the reflection delay wasn't directly centered in a sampled delay bin,
+    # but leaving it here for possible future use b/c it should be much faster.
+    #x = np.array([-1, 0, 1])
+    #a = dfft[np.arange(Ntimes), ref_dly_inds-1]
+    #g = dfft[np.arange(Ntimes), ref_dly_inds+1]
+    #b = dfft[np.arange(Ntimes), ref_dly_inds]
+    #y = np.array([a, b, g]).T
+    #real = np.array([interp1d(x, y[i].real, kind='linear')(bin_shifts[i]) for i in range(Ntimes)])
+    #imag = np.array([interp1d(x, y[i].imag, kind='linear')(bin_shifts[i]) for i in range(Ntimes)])
+    #ref_phs = np.angle(real + 1j*imag)
+
+    # construct reflection
+    if full_freqs is None:
+        full_freqs = freqs
+    eps = construct_reflection(full_freqs, ref_amps, ref_dlys, ref_phs)
+
+    return (eps, ref_amps, ref_dlys, ref_phs, ref_dly_inds, ref_significance,
```

Comment:
I made a mistake here. I was referring to the fact that in real space, if you have a reflection with a total (frequency dependent phase) \theta such that g = 1 + r e^i \theta then |g|^2 = 1 + r^2 + 2 r \cos \theta so the frequency dependent gain term is multiplied by a factor of two and if you fit in real space. However, the cosine term is simply the sum of positive and negative delay components of the reflection which each have an amplitude of |r| so that this issue dissapears if, as you are doing things, the fit occurs in delay space, where the positive/negative delay components are isolated.